File: sdk/typescript/src/client/http-client.ts
import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios';
import { EventEmitter } from 'eventemitter3';
import { CircuitBreaker } from 'circuit-breaker-js';
import retry from 'retry';

import { AuthManager } from '../auth/auth-manager';
import { RequestOptions, SDKError } from '../types';

export interface HttpClientConfig {
  baseURL: string;
  timeout?: number;
  retries?: number;
  retryDelay?: number;
  circuitBreakerOptions?: {
    threshold: number;
    timeout: number;
    resetTimeout: number;
  };
  rateLimit?: {
    requests: number;
    window: number;
  };
  defaultHeaders?: Record<string, string>;
}

export class HttpClient extends EventEmitter {
  private axiosInstance: AxiosInstance;
  private authManager: AuthManager;
  private circuitBreaker: CircuitBreaker;
  private rateLimitQueue: Array<() => void> = [];
  private requestCounts: Map<number, number> = new Map();
  
  constructor(
    config: HttpClientConfig,
    authManager: AuthManager
  ) {
    super();
    
    this.authManager = authManager;
    
    // Create axios instance
    this.axiosInstance = axios.create({
      baseURL: config.baseURL,
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': '@backstage-idp/sdk-typescript/1.0.0',
        ...config.defaultHeaders,
      },
    });

    // Setup circuit breaker
    const cbOptions = config.circuitBreakerOptions || {
      threshold: 5,
      timeout: 60000,
      resetTimeout: 30000
    };
    
    this.circuitBreaker = new CircuitBreaker({
      ...cbOptions,
      onOpen: () => this.emit('circuitBreakerOpen'),
      onHalfOpen: () => this.emit('circuitBreakerHalfOpen'),
      onClose: () => this.emit('circuitBreakerClose'),
    });

    // Setup interceptors
    this.setupRequestInterceptors(config);
    this.setupResponseInterceptors(config);
    
    // Setup rate limiting
    if (config.rateLimit) {
      this.setupRateLimit(config.rateLimit);
    }
  }

  /**
   * Setup request interceptors
   */
  private setupRequestInterceptors(config: HttpClientConfig): void {
    this.axiosInstance.interceptors.request.use(
      async (requestConfig) => {
        // Add authentication header
        try {
          const authHeader = this.authManager.getAuthHeader();
          if (authHeader) {
            if (this.authManager.getAccessToken()?.includes('bearer')) {
              requestConfig.headers['Authorization'] = authHeader;
            } else {
              requestConfig.headers['X-API-Key'] = authHeader;
            }
          }
        } catch (error) {
          // Token refresh failed, continue without auth
        }

        // Add request ID for tracing
        requestConfig.headers['X-Request-ID'] = this.generateRequestId();
        
        this.emit('requestStart', { url: requestConfig.url, method: requestConfig.method });
        
        return requestConfig;
      },
      (error) => {
        this.emit('requestError', error);
        return Promise.reject(this.transformError(error));
      }
    );
  }

  /**
   * Setup response interceptors
   */
  private setupResponseInterceptors(config: HttpClientConfig): void {
    this.axiosInstance.interceptors.response.use(
      (response) => {
        this.emit('requestSuccess', {
          url: response.config.url,
          method: response.config.method,
          status: response.status,
          duration: response.headers['x-response-time']
        });
        
        return response;
      },
      async (error) => {
        const originalRequest = error.config;
        
        // Handle 401 errors with token refresh
        if (error.response?.status === 401 && !originalRequest._retry) {
          originalRequest._retry = true;
          
          try {
            await this.authManager.refreshToken();
            const authHeader = this.authManager.getAuthHeader();
            if (authHeader) {
              if (this.authManager.getAccessToken()?.includes('bearer')) {
                originalRequest.headers['Authorization'] = authHeader;
              } else {
                originalRequest.headers['X-API-Key'] = authHeader;
              }
            }
            
            return this.axiosInstance(originalRequest);
          } catch (refreshError) {
            this.emit('authError', refreshError);
            return Promise.reject(this.transformError(error));
          }
        }

        this.emit('requestError', {
          url: error.config?.url,
          method: error.config?.method,
          status: error.response?.status,
          error: error.message
        });
        
        return Promise.reject(this.transformError(error));
      }
    );
  }

  /**
   * Setup rate limiting
   */
  private setupRateLimit(rateLimit: { requests: number; window: number }): void {
    const windowStart = Math.floor(Date.now() / rateLimit.window);
    this.requestCounts.set(windowStart, 0);

    // Clean up old windows periodically
    setInterval(() => {
      const now = Math.floor(Date.now() / rateLimit.window);
      for (const [window] of this.requestCounts) {
        if (window < now - 10) {
          this.requestCounts.delete(window);
        }
      }
    }, rateLimit.window);
  }

  /**
   * Check rate limit
   */
  private async checkRateLimit(rateLimit?: { requests: number; window: number }): Promise<void> {
    if (!rateLimit) return;

    const now = Date.now();
    const windowStart = Math.floor(now / rateLimit.window);
    const currentCount = this.requestCounts.get(windowStart) || 0;

    if (currentCount >= rateLimit.requests) {
      const waitTime = rateLimit.window - (now % rateLimit.window);
      await new Promise(resolve => setTimeout(resolve, waitTime));
    }

    this.requestCounts.set(windowStart, currentCount + 1);
  }

  /**
   * Make HTTP request with retry and circuit breaker
   */
  public async request<T>(
    config: AxiosRequestConfig,
    options: RequestOptions = {}
  ): Promise<T> {
    return new Promise((resolve, reject) => {
      const operation = retry.operation({
        retries: options.retries || 3,
        factor: 2,
        minTimeout: 1000,
        maxTimeout: 5000,
        randomize: true,
      });

      operation.attempt(async () => {
        try {
          const response = await this.circuitBreaker.execute(async () => {
            return this.axiosInstance.request<T>({
              ...config,
              timeout: options.timeout,
              signal: options.signal,
              headers: {
                ...config.headers,
                ...options.headers,
              },
            });
          });

          resolve(response.data);
        } catch (error) {
          if (operation.retry(error as Error)) {
            return;
          }
          
          reject(this.transformError(error));
        }
      });
    });
  }

  /**
   * GET request
   */
  public async get<T>(
    url: string,
    params?: any,
    options?: RequestOptions
  ): Promise<T> {
    return this.request<T>({
      method: 'GET',
      url,
      params,
    }, options);
  }

  /**
   * POST request
   */
  public async post<T>(
    url: string,
    data?: any,
    options?: RequestOptions
  ): Promise<T> {
    return this.request<T>({
      method: 'POST',
      url,
      data,
    }, options);
  }

  /**
   * PUT request
   */
  public async put<T>(
    url: string,
    data?: any,
    options?: RequestOptions
  ): Promise<T> {
    return this.request<T>({
      method: 'PUT',
      url,
      data,
    }, options);
  }

  /**
   * DELETE request
   */
  public async delete<T>(
    url: string,
    options?: RequestOptions
  ): Promise<T> {
    return this.request<T>({
      method: 'DELETE',
      url,
    }, options);
  }

  /**
   * PATCH request
   */
  public async patch<T>(
    url: string,
    data?: any,
    options?: RequestOptions
  ): Promise<T> {
    return this.request<T>({
      method: 'PATCH',
      url,
      data,
    }, options);
  }

  /**
   * Upload file
   */
  public async upload<T>(
    url: string,
    file: File | Buffer,
    options?: RequestOptions
  ): Promise<T> {
    const formData = new FormData();
    formData.append('file', file);

    return this.request<T>({
      method: 'POST',
      url,
      data: formData,
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    }, options);
  }

  /**
   * Transform axios error to SDK error
   */
  private transformError(error: any): SDKError {
    if (error.isAxiosError) {
      const axiosError = error as AxiosError;
      return {
        code: axiosError.code || 'HTTP_ERROR',
        message: axiosError.message,
        status: axiosError.response?.status,
        details: axiosError.response?.data,
        timestamp: new Date().toISOString(),
      };
    }

    return {
      code: 'UNKNOWN_ERROR',
      message: error.message || 'An unknown error occurred',
      timestamp: new Date().toISOString(),
    };
  }

  /**
   * Generate unique request ID
   */
  private generateRequestId(): string {
    return `req_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Get circuit breaker status
   */
  public getCircuitBreakerStatus(): {
    state: 'closed' | 'open' | 'half-open';
    failures: number;
    nextAttempt?: number;
  } {
    return {
      state: this.circuitBreaker.state,
      failures: this.circuitBreaker.failures,
      nextAttempt: this.circuitBreaker.nextAttempt,
    };
  }

  /**
   * Reset circuit breaker
   */
  public resetCircuitBreaker(): void {
    this.circuitBreaker.reset();
  }

  /**
   * Dispose of the client
   */
  public dispose(): void {
    this.removeAllListeners();
  }
}

File: sdk/python/backstage_idp_sdk/client/http_client.py
"""HTTP client for Backstage IDP SDK."""

import asyncio
import time
from typing import Any, Dict, Optional, Union, List
from urllib.parse import urljoin

import httpx
from tenacity import (
    retry, 
    stop_after_attempt, 
    wait_exponential,
    retry_if_exception_type,
    RetryError as TenacityRetryError
)

from ..auth.auth_manager import AuthManager
from ..types import ClientConfig, RequestOptions
from ..exceptions import (
    BackstageSDKError,
    NetworkError,
    TimeoutError,
    CircuitBreakerError,
    create_error_from_response,
    is_retryable_error
)


class CircuitBreaker:
    """Simple circuit breaker implementation."""
    
    def __init__(self, failure_threshold: int = 5, timeout: float = 60.0, recovery_timeout: float = 30.0):
        self.failure_threshold = failure_threshold
        self.timeout = timeout
        self.recovery_timeout = recovery_timeout
        self.failure_count = 0
        self.last_failure_time = 0
        self.state = "closed"  # closed, open, half_open

    async def call(self, func, *args, **kwargs):
        """Execute function with circuit breaker protection."""
        if self.state == "open":
            if time.time() - self.last_failure_time > self.timeout:
                self.state = "half_open"
            else:
                raise CircuitBreakerError("Circuit breaker is open")

        try:
            result = await func(*args, **kwargs)
            if self.state == "half_open":
                self.reset()
            return result
        except Exception as e:
            self.record_failure()
            raise e

    def record_failure(self):
        """Record a failure."""
        self.failure_count += 1
        self.last_failure_time = time.time()
        
        if self.failure_count >= self.failure_threshold:
            self.state = "open"

    def reset(self):
        """Reset circuit breaker."""
        self.failure_count = 0
        self.state = "closed"
        self.last_failure_time = 0


class RateLimiter:
    """Token bucket rate limiter."""
    
    def __init__(self, requests_per_second: float, burst_size: int):
        self.requests_per_second = requests_per_second
        self.burst_size = burst_size
        self.tokens = burst_size
        self.last_update = time.time()
        self.lock = asyncio.Lock()

    async def acquire(self):
        """Acquire a token for making a request."""
        async with self.lock:
            now = time.time()
            # Add tokens based on elapsed time
            elapsed = now - self.last_update
            self.tokens = min(
                self.burst_size, 
                self.tokens + elapsed * self.requests_per_second
            )
            self.last_update = now
            
            if self.tokens >= 1:
                self.tokens -= 1
                return
            
            # Wait for next token
            wait_time = (1 - self.tokens) / self.requests_per_second
            await asyncio.sleep(wait_time)
            self.tokens = 0


class HttpClient:
    """HTTP client with retry, circuit breaker, and rate limiting."""
    
    def __init__(self, config: ClientConfig, auth_manager: Optional[AuthManager] = None):
        self.config = config
        self.auth_manager = auth_manager
        
        # Initialize HTTP client
        self.client = httpx.AsyncClient(
            timeout=httpx.Timeout(config.timeout or 30.0),
            verify=config.verify_ssl,
            headers={
                "User-Agent": "backstage-idp-sdk-python/1.0.0",
                "Accept": "application/json",
                "Content-Type": "application/json"
            }
        )
        
        # Initialize circuit breaker
        self.circuit_breaker = CircuitBreaker()
        
        # Initialize rate limiter if configured
        self.rate_limiter: Optional[RateLimiter] = None
        if hasattr(config, 'rate_limit') and config.rate_limit:
            self.rate_limiter = RateLimiter(
                requests_per_second=config.rate_limit['requests'] / 60.0,  # Convert per minute to per second
                burst_size=min(config.rate_limit['requests'], 10)
            )

    async def request(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        json: Optional[Any] = None,
        data: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        options: Optional[RequestOptions] = None
    ) -> Any:
        """Make HTTP request with retry and error handling."""
        
        # Apply rate limiting
        if self.rate_limiter:
            await self.rate_limiter.acquire()
            
        # Prepare request
        url = urljoin(self.config.base_url, path.lstrip('/'))
        request_headers = dict(self.client.headers)
        
        # Add auth header
        if self.auth_manager:
            try:
                auth_header = self.auth_manager.get_auth_header()
                if auth_header:
                    if auth_header.startswith('Bearer'):
                        request_headers['Authorization'] = auth_header
                    else:
                        request_headers['X-API-Key'] = auth_header
            except Exception:
                # Continue without auth if token refresh fails
                pass
                
        # Add custom headers
        if headers:
            request_headers.update(headers)
            
        # Add request ID for tracing
        request_headers['X-Request-ID'] = self._generate_request_id()
        
        # Configure retry
        retry_decorator = retry(
            stop=stop_after_attempt(self.config.retries or 3),
            wait=wait_exponential(
                multiplier=self.config.retry_delay or 1.0,
                max=30
            ),
            retry=retry_if_exception_type((httpx.RequestError, httpx.TimeoutException)),
            reraise=True
        )
        
        @retry_decorator
        async def _make_request():
            return await self.circuit_breaker.call(
                self._execute_request,
                method=method,
                url=url,
                params=params,
                json=json,
                data=data,
                headers=request_headers,
                timeout=options.timeout if options else None
            )
            
        try:
            return await _make_request()
        except TenacityRetryError as e:
            # Get the original exception
            original_error = e.last_attempt.exception()
            raise self._transform_error(original_error)
        except Exception as e:
            raise self._transform_error(e)

    async def _execute_request(
        self,
        method: str,
        url: str,
        params: Optional[Dict[str, Any]] = None,
        json: Optional[Any] = None,
        data: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        timeout: Optional[float] = None
    ) -> Any:
        """Execute the actual HTTP request."""
        
        timeout_config = httpx.Timeout(timeout) if timeout else None
        
        try:
            response = await self.client.request(
                method=method,
                url=url,
                params=params,
                json=json,
                data=data,
                headers=headers,
                timeout=timeout_config
            )
            
            # Handle authentication errors with token refresh
            if response.status_code == 401 and self.auth_manager:
                try:
                    await self.auth_manager.refresh_token()
                    # Retry request with new token
                    auth_header = self.auth_manager.get_auth_header()
                    if auth_header:
                        if auth_header.startswith('Bearer'):
                            headers = headers or {}
                            headers['Authorization'] = auth_header
                        else:
                            headers = headers or {}
                            headers['X-API-Key'] = auth_header
                            
                    response = await self.client.request(
                        method=method,
                        url=url,
                        params=params,
                        json=json,
                        data=data,
                        headers=headers,
                        timeout=timeout_config
                    )
                except Exception:
                    # If refresh fails, continue with original response
                    pass
            
            # Check for error responses
            if response.status_code >= 400:
                try:
                    error_data = response.json()
                except Exception:
                    error_data = {
                        'error': 'HTTP_ERROR',
                        'message': f'HTTP {response.status_code}: {response.reason_phrase}'
                    }
                
                raise create_error_from_response(error_data, response.status_code)
            
            # Return JSON response or empty dict for 204 No Content
            if response.status_code == 204:
                return {}
                
            try:
                return response.json()
            except Exception:
                # Return text content if not JSON
                return response.text
                
        except httpx.TimeoutException as e:
            raise TimeoutError(f"Request timeout: {e}")
        except httpx.RequestError as e:
            raise NetworkError(f"Network error: {e}")

    async def get(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        options: Optional[RequestOptions] = None
    ) -> Any:
        """Make GET request."""
        return await self.request("GET", path, params=params, headers=headers, options=options)

    async def post(
        self,
        path: str,
        json: Optional[Any] = None,
        data: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        options: Optional[RequestOptions] = None
    ) -> Any:
        """Make POST request."""
        return await self.request("POST", path, json=json, data=data, headers=headers, options=options)

    async def put(
        self,
        path: str,
        json: Optional[Any] = None,
        data: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        options: Optional[RequestOptions] = None
    ) -> Any:
        """Make PUT request."""
        return await self.request("PUT", path, json=json, data=data, headers=headers, options=options)

    async def patch(
        self,
        path: str,
        json: Optional[Any] = None,
        data: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        options: Optional[RequestOptions] = None
    ) -> Any:
        """Make PATCH request."""
        return await self.request("PATCH", path, json=json, data=data, headers=headers, options=options)

    async def delete(
        self,
        path: str,
        headers: Optional[Dict[str, str]] = None,
        options: Optional[RequestOptions] = None
    ) -> Any:
        """Make DELETE request."""
        return await self.request("DELETE", path, headers=headers, options=options)

    def get_circuit_breaker_status(self) -> Dict[str, Any]:
        """Get circuit breaker status."""
        return {
            "state": self.circuit_breaker.state,
            "failures": self.circuit_breaker.failure_count,
            "last_failure": self.circuit_breaker.last_failure_time
        }

    def reset_circuit_breaker(self):
        """Reset circuit breaker."""
        self.circuit_breaker.reset()

    def _generate_request_id(self) -> str:
        """Generate unique request ID."""
        return f"req_{int(time.time() * 1000)}_{id(self) % 10000:04d}"

    def _transform_error(self, error: Exception) -> BackstageSDKError:
        """Transform exception to SDK error."""
        if isinstance(error, BackstageSDKError):
            return error
            
        if isinstance(error, httpx.TimeoutException):
            return TimeoutError(f"Request timeout: {error}")
        elif isinstance(error, httpx.RequestError):
            return NetworkError(f"Network error: {error}")
        else:
            return BackstageSDKError(f"Unexpected error: {error}")

    async def close(self):
        """Close the HTTP client."""
        await self.client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()